src/nvim/eval.c

```diff
@@ -21032,6 +21032,23 @@ void ex_oldfiles(exarg_T *eap)
     /* Assume "got_int" was set to truncate the listing. */
     got_int = FALSE;
 
+    // File selection prompt on ":oldfiles!"
+    if (eap->forceit) {
+      quit_more = false;
+      nr = prompt_for_number(false);
+      msg_starthere();
+      if (nr > 0) {
+        char_u *p = list_find_str(get_vim_var_list(VV_OLDFILES), nr);
```

Comment:
`get_vim_var_list()` basically gets me `v:oldfiles` and `list_find_str()` gets me the correct entry.